tests/mitmproxy_addons/README.md

```diff
@@ -21,17 +18,99 @@ POST /options/lock
  {
    "options": {
      "body_size_limit": "3m",
+     "callback": {
+       "callback_response_url": "http://host.docker.internal:445566"
+     }
    }
  }
  HTTP/1.1 200 OK
  {
    "reset_id": "some_opaque_string"
  }
 ```
+Any [option](https://docs.mitmproxy.org/stable/concepts-options/) can be specified in the
+`options` object, not just Complement specific addons.
 
 ```
 POST /options/unlock
 {
    "reset_id": "some_opaque_string"
 }
-```
\ No newline at end of file
+```
+
+Tests will lock/unlock whenever they need to interact with mitmproxy. Attempting to lock an already locked controller will return an HTTP 400 error. Attempting to unlock an already unlocked controller will return an HTTP 400 error.
+
+### Callback addon
+
+A [mitmproxy addon](https://docs.mitmproxy.org/stable/addons-examples/) bolts on custom
+functionality to mitmproxy. This typically involves using the
+[Event Hooks API](https://docs.mitmproxy.org/stable/api/events.html) to listen for
+[HTTP flows](https://docs.mitmproxy.org/stable/api/mitmproxy/http.html#HTTPFlow).
+
+The `callback` addon is a Complement-Crypto specific addon which calls a client provided URL
+mid-flow, with a JSON object containing information about the HTTP flow. The caller can then
+return another JSON object which can modify the response in some way.
+
+Available configuration options are optional:
+ - `callback_request_url`: the URL to send outbound requests to. This allows callbacks to intercept
+   requests BEFORE they reach the server.
+ - `callback_response_url`: the URL to send inbound responses to. This allows callbacks to modify
+   response content.
+ - `filter`: the [mitmproxy filter](https://docs.mitmproxy.org/stable/concepts-filters/) to apply. If unset, ALL requests are eligible to go to the callback
+   server.
+
+To use this with the controller API, you would send an HTTP request like this:
+```js
+{
+  "options": {
+    "callback": {
+      "callback_response_url": "http://host.docker.internal:445566/response"
+    }
+  }
+}
+```
+
+#### `callback_response_url`
+mitmproxy will POST to `callback_response_url` with the following JSON object:
+```js
+{
+   method: "GET|PUT|...",
+   access_token: "syt_11...",
+   url: "http://hs1/_matrix/client/...",
+   request_body: { some json object or null if no body },
+   response_body: { some json object },
```

Comment:
are they necessarily JSON? What if, say, the intercepted request was for a piece of media?